lib/object.c

```diff
@@ -5,11 +5,178 @@
+                .details = {}
+             },
+             .keyBits = 2048,
+             .exponent = 0,
+         },
+        .unique.rsa = {
+             .size = 0,
+         }
+     }
+};
+
+static const TPM2B_PUBLIC primary_ecc_template = {
+    .publicArea = {
+        .type = TPM2_ALG_ECC,
+        .nameAlg = TPM2_ALG_SHA256,
+        .objectAttributes = (TPMA_OBJECT_USERWITHAUTH |
+                             TPMA_OBJECT_RESTRICTED |
+                             TPMA_OBJECT_DECRYPT |
+                             TPMA_OBJECT_NODA |
+                             TPMA_OBJECT_FIXEDTPM |
+                             TPMA_OBJECT_FIXEDPARENT |
+                             TPMA_OBJECT_SENSITIVEDATAORIGIN),
+        .authPolicy = {
+             .size = 0,
+         },
+        .parameters.eccDetail = {
+             .symmetric = {
+                 .algorithm = TPM2_ALG_AES,
+                 .keyBits.aes = 128,
+                 .mode.aes = TPM2_ALG_CFB,
+              },
+             .scheme = {
+                .scheme = TPM2_ALG_NULL,
+                .details = {}
+             },
+             .curveID = TPM2_ECC_NIST_P256,
+             .kdf = {
+                .scheme = TPM2_ALG_NULL,
+                .details = {}
+             },
+         },
+        .unique.ecc = {
+             .x.size = 0,
+             .y.size = 0
+         }
+     }
+};
+tool_rc setup_primary(ESYS_CONTEXT *esys_ctx, ESYS_TR *parent) {
+    tool_rc rc;
+    TSS2_RC r;
+    const TPM2B_PUBLIC *primary_template = NULL;
+    TPMS_CAPABILITY_DATA *capability_data = NULL;
+
+    r = Esys_GetCapability(esys_ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+                            TPM2_CAP_ALGS, 0, TPM2_MAX_CAP_ALGS, NULL,
+                            &capability_data);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_ERR("Unable to fetch TPM supported algorithms");
+        rc = tool_rc_general_error;
+        goto ret;
+    }
+
+    for (uint32_t index = 0; index < capability_data->data.algorithms.count; index++) {
+        if (capability_data->data.algorithms.algProperties[index].alg == TPM2_ALG_ECC) {
+            primary_template = &primary_ecc_template;
+            break;
+        }
+    }
+    if (primary_template == NULL) {
+        primary_template = &primary_rsa_template;
+    }
+
+    if(!primary_template) {
+        LOG_ERR("Primary Template not correctly populated");
+        rc = tool_rc_general_error;
+        goto ret;
+    }
+
+    r = Esys_CreatePrimary(esys_ctx, ESYS_TR_RH_OWNER,
+                            ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+                            &primary_sensitive, primary_template, &all_outside_info,
+                            &all_creation_PCR,
+                            parent, NULL, NULL, NULL, NULL);
+
+    if (r != TSS2_RC_SUCCESS) {
+        rc = tool_rc_general_error;
+        goto ret;
+    }
+    rc = tool_rc_success;
+
+ret:
+    if (capability_data)
+        free(capability_data);
```

Comment:
capability_data is allocated space in "Esys_GetCapability" somewhere, so freeing that.

Reply:
Use Esys_Free() and no need to check if set, NULL is OK to pass.

Reply:
Anything allocated from Esys needs to be free'd with Esys_Free or it breaks on windows where the library and tool could be linked against different implementations on malloc, so if someone malloc's with malloc1 and free's with free2 or vice versa it blows up.